Add knob for Vertex Buffer Format

elviscapiaq opened the pull request:
- It was added a new Vertex Buffer that stores 16-bit floating point format for the vertex input including position, texCoord, normal and tangent.

wangra-google commented on benchmarks/graphics_pipeline/main.cpp:
```diff
@@ -379,8 +404,13 @@ void ProjApp::Setup()
-        PPX_CHECKED_CALL(Geometry::Create(GeometryOptions::InterleavedU32().AddTexCoord().AddNormal().AddTangent(), &geo));
+        Geometry        lowPrecision;
+        GeometryOptions options;
+        options.IndexTypeU32().AddPosition(grfx::FORMAT_R16G16B16_FLOAT).AddTexCoord(grfx::FORMAT_R16G16_FLOAT).AddNormal(grfx::FORMAT_R16G16B16_FLOAT).AddTangent(grfx::FORMAT_R16G16B16A16_FLOAT);
```
i think for normal and binormal, FORMAT_R8G8B8A8_SNORM should be good enough

elviscapiaq replied:
Updated.

wangra-google commented on include/ppx/tri_mesh.h:
```diff
@@ -60,6 +60,19 @@ struct TriMeshVertexData
+//! @struct TriMesh16VertexData16
+//!
+//!
+struct TriMeshVertexData16
```
Probably better with the name `TriMeshVertexDataCompressed`

elviscapiaq replied:
Done.

wangra-google commented on include/ppx/tri_mesh.h:
```diff
@@ -60,6 +60,19 @@ struct TriMeshVertexData
+//!
+struct TriMeshVertexData16
+{
+    glm::u16vec3 position;
```
i think to be consistent with the other vector types, it would be better to do something like `using half3 = glm::u16vec3`

elviscapiaq replied:
Done.

wangra-google commented on src/ppx/geometry.cpp:
```diff
@@ -247,6 +260,39 @@ class VertexDataProcessorInterleaved : public VertexDataProcessorBase
 
+    virtual uint32_t AppendVertexData(Geometry* pGeom, const TriMeshVertexData16& vtx) override
+    {
+        uint32_t       startSize = GetVertexBufferSize(pGeom, kBufferIndex);
```
can this code block within the function being templated instead of a copy of existing code?

elviscapiaq replied:
Thanks for the feedback. Updated.

wangra-google commented on src/ppx/geometry.cpp:
```diff
@@ -363,6 +409,36 @@ class VertexDataProcessorPositionPlanar : public VertexDataProcessorBase
 
+    virtual uint32_t AppendVertexData(Geometry* pGeom, const TriMeshVertexData16& vtx) override
+    {
+        const uint32_t n = AppendDataToVertexBuffer(pGeom, GetPositionBufferIndex(pGeom), vtx.position);
```
can this code block within the function being templated instead of a copy of existing code?

elviscapiaq replied:
Updated.

wangra-google commented on src/ppx/geometry.cpp:
```diff
@@ -177,6 +178,18 @@ class VertexDataProcessorPlanar : public VertexDataProcessorBase
 
+    virtual uint32_t AppendVertexData(Geometry* pGeom, const TriMeshVertexData16& vtx) override
+    {
+        const uint32_t n = AppendDataToVertexBuffer(pGeom, GetPositionBufferIndex(pGeom), vtx.position);
```
can this code block within the function being templated instead of a copy of existing code?

elviscapiaq replied:
Updated.

elviscapiaq commented:
I updated the code such that the function `AppendVertexData `is being templated instead of copying code.
I had an issue here since the class templated function `AppendVertexData ` implements a pure virtual function in the derived classes. I found a solution to address this issue. Let me know if this make sense. Ready for review.

wangra-google commented on benchmarks/graphics_pipeline/main.cpp:
```diff
@@ -379,8 +404,14 @@ void ProjApp::Setup()
+        Geometry        lowPrecision;
+        GeometryOptions options;
+        options.isVertexDataCompressed = true;
+        options.IndexTypeU32().AddPosition(grfx::FORMAT_R16G16B16_FLOAT).AddTexCoord(grfx::FORMAT_R16G16_FLOAT).AddNormal(grfx::FORMAT_R8G8B8_SNORM).AddTangent(grfx::FORMAT_R8G8B8A8_SNORM);
```
is there a reason why you use rgb8 for normal but rgba8 for tangent? normally rgba8 should be fine. better to be aligned to 4 bytes

elviscapiaq replied:
Updated.

elviscapiaq commented:
I fixed the code such that the floats are mapped to a int8. It seems when we use the `FORMAT_R8G8B8A8_SNORM` , the shader maps values (`int8` type) from `[-128, 127]` to `[-1, 1]`, so I added a function to map a float between [-1, 1] to [-128, 127]. It should work because normal and tangents have all component values between [-1, 1]. Ready for review.

elviscapiaq commented:
I rebased this PR. I think it is ready to be merged.

Keenuts commented:
> I rebased this PR. I think it is ready to be merged.

This PR doesn't compile for me (and the CI do break, so reassuring 😊 )

elviscapiaq commented:
I fixed the issue producing the tests failed and replaced the function to convert a float to a 16-bits IEEE-754 with the glm version to avoid license issues. Ready to be merged.

Keenuts commented on benchmarks/graphics_pipeline/main.cpp:
```diff
@@ -844,10 +882,11 @@ void ProjApp::Render()
-            frame.cmd->BindIndexBuffer(mSphere.mesh);
-            frame.cmd->BindVertexBuffers(mSphere.mesh);
+            const size_t vbFormatIndex = pKnobVbFormat->GetIndex();
+            const size_t pipelineIndex = pKnobVs->GetIndex() * kAvailablePsShaders.size() * kAvailableVbFormats.size() + pKnobPs->GetIndex() * kAvailableVbFormats.size() + vbFormatIndex;
```
"This is getting out of hand!" 🙃
We might want to revisit this in another PR. Keeping adding new indices and pipelines in the lest and computing the index depending on what we need is getting messy.
